### PR TITLE
Transforming ip and op strings into Sets and change restrictions type

### PR DESCRIPTION
**Pull request by rafaelgmota**
With input and output ports type modification, we get ride of a lot of string manipulation and make the code more readable.

With restrictions type modification, we simplify addRestriction and getRestriction methods

### CODE REVIEW

File: src/rbr/Region.java [hunk 1]
```diff
@@ -6,12 +6,12 @@
 public class Region {
 	// @ToDo inputPortsSet and outputPortsSet could actually be Sets. This would ease many operations on them:
```
Comment (rmilfont) [inline src/rbr/Region.java]: Remove these comment lines.

File: src/util/GraphRestrictions.java [hunk 2]
```diff
@@ -1,61 +1,40 @@
 package util;
 
-import java.util.HashMap;
+import java.util.*;
 
 public class GraphRestrictions {
-    private HashMap<Vertex, String[]> restrictions;
+    private Map<Vertex, Map<Character, Set<Character>>> restrictions;
+    //
```
Comment (rmilfont) [inline src/util/GraphRestrictions.java]: Remove this line.

File: src/rbr/RBR.java [hunk 3]
```diff
@@ -271,27 +276,26 @@ private double reachability(Vertex orig) {
...
 		if (opColor == null)
 			return false;
 		return reaches(src.adjunct(opColor).destination(), dest, EdgeColor.getInvColor(src.adjunct(opColor).color()));
 	}
 
-	private String getOpColor(Vertex src, Vertex dest, String ipColor) {
+	private Character getOpColor(Vertex src, Vertex dest, Character ipColor) {
 		int destX = Integer.valueOf(dest.name().split("\\.")[0]);
 		int destY = Integer.valueOf(dest.name().split("\\.")[1]);
 		for (rbr.Region reg : regionsForVertex.get(src))
```
Comment (rmilfont) [inline src/rbr/RBR.java]: Put the braces for 'for' statement.

File: src/rbr/Region.java [hunk 4]
```diff
@@ -65,34 +65,35 @@ public String toString() {
...
 		return op;
 	}
 
-	private static String getIpMerged(Region r1, Region r2) {
-		String ip = r2.inputPorts();
-		for (int i = 0; i < r1.inputPorts().length(); i++) {
-			if (!ip.contains(r1.inputPorts().substring(i, i + 1)))
-				ip += r1.inputPorts().substring(i, i + 1);
+	private static Set<Character> getIpMerged(Region r1, Region r2) {
+		Set<Character> ipr1 = r1.inputPorts();
```
Comment (rmilfont) [inline src/rbr/Region.java]: Rename 'ipr' variables for a more comprehensive name

File: src/rbr/Region.java [hunk 5]
```diff
@@ -113,14 +114,8 @@ boolean canBeMergedWith(Region that) {
...
-		String r1Op = sortStrAlf(r1.outputPorts());
-		String r2Op = sortStrAlf(r2.outputPorts());
-		return (r1Op.contains(r2Op) || r2Op.contains(r1Op));
-	}
-
-	private static String sortStrAlf(String input) {
-		char[] ip1 = input.toCharArray();
-		Arrays.sort(ip1);
-		return String.valueOf(ip1);
+		Set<Character> r1Op = r1.outputPorts();
```
Comment (rmilfont) [inline src/rbr/Region.java]: Rename 'rNOp' variables for a more comprehensive name

File: src/rbr/RoutingTableGenerator.java [hunk 6]
```diff
@@ -7,8 +7,7 @@
 import java.io.File;
 import java.io.FileWriter;
 import java.io.IOException;
```
Comment (rmilfont) [inline src/rbr/RoutingTableGenerator.java]: Squash the imports using '*'.

**Review by rmilfont — CHANGES_REQUESTED** [review]

**Comment by rafaelgmota** [comment]
Done above modifications and get ride of getIpMerged and getOpMerged

File: src/rbr/RBR.java [hunk 7]
```diff
@@ -25,14 +25,14 @@ private void packOutputPort(Vertex atual) {
...
+			Set<Character> op = a.getOp();
+			Set<Character> ip = a.getIp();
 			Vertex dst = a.destination();
-			String ip = a.getIp();
 
 			for (RoutingOption b : actRP) {
 				if (ip.equals(b.getIp()) && dst.equals(b.destination())) {
-					if (!op.contains(b.getOp()))
-						op = op.concat(b.getOp());
+					if (!op.containsAll(b.getOp()))
```
Comment (jmferreiratech) [inline src/rbr/RBR.java]: You don't need to check it, do you?

File: src/paths_gen/PathFinder.java [hunk 8]
```diff
@@ -61,7 +62,7 @@ public PathFinder(Graph g, GraphRestrictions restrictions){
         ArrayList<Path> result = new ArrayList<>();
         Vertex currentSrc = p.dst();
         Vertex predecessor = p.get(p.size() - 2);
-        String inputPort = currentSrc.edge(predecessor).color();
+        char inputPort = currentSrc.edge(predecessor).color();
```
Comment (jmferreiratech) [inline src/paths_gen/PathFinder.java]: Use a Character.

File: src/rbr/RBR.java [hunk 9]
```diff
@@ -45,14 +45,15 @@ public void packInputPort(Vertex atual) {
...
+			Set<Character> op = a.getOp();
+			Set<Character> ip = a.getIp();
 			Vertex dst = a.destination();
-			String ip = a.getIp();
 
 			for (RoutingOption b : actRP) {
 				if (op.equals(b.getOp()) && dst.equals(b.destination())) {
-					if (!ip.contains(b.getIp()))
-						ip = ip.concat(b.getIp());
+					if (!ip.containsAll(b.getIp())) {
```
Comment (jmferreiratech) [inline src/rbr/RBR.java]: You don't need to check it, do you?

File: src/rbr/RBR.java [hunk 10]
```diff
@@ -238,9 +244,8 @@ private static Range box(Set<Vertex> dsts) {
 		return result;
 	}
 
-	private Region makeRegion(int xmin, int ymin, int xmax, int ymax,
-			String ip, String op) {
-		Set<Vertex> dst = new HashSet<>();
+	private Region makeRegion(int xmin, int ymin, int xmax, int ymax, Set<Character> ip, Set<Character> op) {
+			Set<Vertex> dst = new HashSet<>();
```
Comment (jmferreiratech) [inline src/rbr/RBR.java]: This line indentation looks weird on github.

File: src/rbr/RBR.java [hunk 11]
```diff
@@ -271,27 +277,29 @@ private double reachability(Vertex orig) {
...
 		int destY = Integer.valueOf(dest.name().split("\\.")[1]);
-		for (rbr.Region reg : regionsForVertex.get(src))
-			if (reg.box().contains(destX, destY) && reg.inputPorts().contains(ipColor))
-				return (reg.outputPorts().substring(0, 1));
-
-		System.err.println("ERROR : There isn't Op on " + src.name()
-				+ "("  + ipColor + ") going to " + dest.name());
+		for (rbr.Region reg : regionsForVertex.get(src)) {
+			if (reg.box().contains(destX, destY) && reg.inputPorts().contains(ipColor)) {
+				return (Character)reg.outputPorts().toArray()[0];
```
Comment (jmferreiratech) [inline src/rbr/RBR.java:296]: You really need this cast?

File: src/rbr/Region.java [hunk 12]
```diff
@@ -4,14 +4,12 @@
...
+	private Set<Character> inputPortsSet;
+	private Set<Character> outputPortsSet;
 	private Range box;
 	private Set<Vertex> destinations;
 
-	public Region(String ip, Set<Vertex> destinations, String op) {
+	public Region(Set<Character> ip, Set<Vertex> destinations, Set<Character> op) {
 		this.destinations = new HashSet<>(destinations);
 		this.inputPortsSet = ip;
 		this.outputPortsSet = op;
```
Comment (jmferreiratech) [inline src/rbr/Region.java]: Maybe is more safe to make a copy here as these sets could be changed by the caller.

File: src/rbr/Region.java [hunk 13]
```diff
@@ -4,14 +4,12 @@
 import util.*;
 
 public class Region {
-	// @ToDo inputPortsSet and outputPortsSet could actually be Sets. This would ease many operations on them:
-	// @ToDo isSubset, getOpMerged (intersection), getIpMerged (union) and so on.
-	private String inputPortsSet;
-	private String outputPortsSet;
+	private Set<Character> inputPortsSet;
+	private Set<Character> outputPortsSet;
```
Comment (jmferreiratech) [inline src/rbr/Region.java]: Rename those variables: remove the Set terminator.

File: src/rbr/Region.java [hunk 14]
```diff
@@ -109,18 +88,7 @@ private static String getIpMerged(Region r1, Region r2) {
...
-		String r2Op = sortStrAlf(r2.outputPorts());
-		return (r1Op.contains(r2Op) || r2Op.contains(r1Op));
-	}
-
-	private static String sortStrAlf(String input) {
-		char[] ip1 = input.toCharArray();
-		Arrays.sort(ip1);
-		return String.valueOf(ip1);
+		return (this.box().isContiguous(that.box()) && (this.outputPorts().containsAll(that.outputPorts())
+				|| that.outputPorts().containsAll(this.outputPorts())));
```
Comment (jmferreiratech) [inline src/rbr/Region.java]: I'd rather keep the opIsSubset method with these two conditions. Maybe changing its name.

File: src/sbr/SR.java [hunk 15]
```diff
@@ -314,54 +314,58 @@ public GraphRestrictions restrictions() {
...
+				oPStarting.add(Starting.edge(Ending).color());
+				Set<Character> oPEnding = new HashSet<>();
+				oPEnding.add(Ending.edge(Starting).color());
 				// Restrictions at Starting core
-				for (Edge link : Starting.adjuncts())
-					if (link.color() != opStarting) {
-						restrictions.addRestriction(Starting, link.color(), opStarting);
-					}
+				for (Edge link : Starting.adjuncts()) {
+						restrictions.addRestriction(Starting, link.color(), oPStarting);
```
Comment (jmferreiratech) [inline src/sbr/SR.java]: addRestriction always receive only one restriction, so you don't need to receive a Set for 'ip' and 'op' (if you needed, you should call it addRestrictions). Making it receive a Character will simplify the setrestrictions method.

**Review by jmferreiratech — CHANGES_REQUESTED** [review]
Good job! Make these few mods and I'll upstream it.

**Comment by rafaelgmota** [comment]
Above modifications were done

**Review by jmferreiratech — APPROVED** [review]

**Pull request merged**